Adding third party module - example (#73)
moondance-labs/tanssi-docs: builders/build/customize/adding-external-module.md

Similarly to what is described in the [built-in module](/builders/build/customize/adding-built-in-module/#adding-a-built-in-module-to-runtime){target=_blank} article, adding a third-party module requires the following steps:

1. Declare the Dependency
2. Make the Standard Features Available to the Compiler
3. Configure and add the module to the Runtime

Should the third-party module reference any dependency already referenced from a distinct source or version, compilation will fail. To resolve this issue, it will be necessary to apply a patch.

#### Declaring the Dependency {: #declaring-dependency }

Declare in the `Cargo.toml` located in the repository's root folder the [toggle module](https://github.com/papermoonio/pallet-toggle.git){target=_blank} under the section `[dependencies]`. 

This `toggle` module, built for testing and educational purposes, adds a basic logic to the runtime, allowing users to switch a state between true and false.

```toml
[dependencies]
...
pallet-toggle = { 
    git = "https://github.com/papermoonio/pallet-toggle.git", 
    default-features = false 
}
...
```

#### Make the Standard Features Available to the Compiler {: #add-standard-features }

> Having declared the module in the workspace `Cargo.toml` file, the dependency can now be added to the specific template `Cargo.toml` file, which is located in the folder `container-chains/templates/simple/runtime`.

Reviewer: ```suggestion
Having declared the module in the workspace `Cargo.toml` file, the dependency can now be added to the specific template `Cargo.toml` file, which for this example that uses the Tanssi GitHub repo, is located in the folder `container-chains/templates/simple/runtime`.
```

```toml
[dependencies]
...
pallet-toggle = { workspace = true }
...
```

In the same `Cargo.toml` file, add the following features.

```toml
[features]
default = [
	"std",
]
std = [
	...,
	"pallet-toggle/std",
   ...
]
...
runtime-benchmarks = [
	...,
	"pallet-toggle/runtime-benchmarks",
]

try-runtime = [
	...,
	"pallet-toggle/try-runtime",
]
```

> #### Configure and Add the Module to the Runtime {: #configure-module-in-the-runtime }

Reviewer: ```suggestion
### Configure and Add the Module to the Runtime {: #configure-module-in-the-runtime }
```

> Add the following snippet that configures the module and add the module within the `construct_runtime!`macro.

Reviewer: ```suggestion
Next, add the following snippet in the `lib.rs` file inside the runtime folder. This configures the module and add the module within the `construct_runtime!`macro.
```

```rust
...
impl pallet_template::Config for Runtime {
	type RuntimeEvent = RuntimeEvent;
	type WeightInfo = pallet_template::weights::SubstrateWeight<Runtime>;
}

construct_runtime!(
    pub enum Runtime
    {
        ...
        ...
        Toggle: pallet_toggle,
    }
);
```

> ### Compile {: #compile }

Reviewer: ```suggestion
### Compile Runtime {: #compile-runtime }
```

After completing the preceding steps, the module is declared a dependency in the project, configured, and added to the runtime. 

Compile the template using the following command:

```bash
cargo build -p container-chain-template-simple-node --release
```

The terminal output will display an error, similar to the following, caused by different modules referencing different versions of the same dependency:

```bash
error: failed to select a version for `syn`.
```

### Patch depedencies {: #patch-dependencies }

> Finally, executing the `diener` [command](#solving-dependencies-conflicts-diener) will add a `patch` section to your workspace `Cargo.toml` file, overriding the dependencies and unifying origins and versions.

Reviewer: Add an example of how the diener command is executed for this example

This command applies the changes to the `Cargo.toml` file, patching the dependencies, and solving the double reference issues.

You can visit the [diener documentation](https://docs.rs/crate/diener/latest){target=_blank} to learn more about the tool and other extra functions it offers.

## Example of the Double Reference Issue {: #double-reference-issue }

> To illustrate the situation, the following steps add a demo [external module](https://github.com/papermoonio/pallet-toggle.git){target=_blank} to a custom runtime based on the [baseline Appchain template](/builders/build/templates/substrate/){target=_blank}. One way to follow this tutorial is to clone the [Tanssi Github repository](https://github.com/moondance-labs/tanssi){target=_blank}, which will act as the root repository of the project.

Reviewer 2: All of the `target=_blank`s on this page should be changed to `target=\_blank`

1. Declare the dependency in the root `Cargo.toml` file
2. Make the standard features available to the compiler
3. Configure and add the module to the runtime

Should the third-party module reference any dependency already referenced from a distinct source or version, compilation will fail. 

The following diagram shows how two different references to the same dependency are being included in the runtime, causing the compilation to fail:

![Double reference](/images/builders/build/external-module/external-module-1.png)

To resolve this issue, it will be necessary to apply a patch so that the references for the dependency are unified:

![Patched reference](/images/builders/build/external-module/external-module-2.png)

### Declaring the Dependency {: #declaring-dependency }

The first step to reproduce the double reference issue is to declare the dependency in the `Cargo.toml` file located in the repository's root folder, under the section `[dependencies]`. For this example, a simple [toggle module](https://github.com/papermoonio/pallet-toggle.git){target=_blank} is used. 

> This `toggle` module, built for testing and educational purposes, adds a basic logic to the runtime, allowing users to switch a state between true and false.

Reviewer 2: ```suggestion
This `toggle` module, built for testing and educational purposes, adds basic logic to the runtime, allowing users to switch a state between true and false.
```

```toml
[dependencies]
...
pallet-toggle = { 
    git = "https://github.com/papermoonio/pallet-toggle.git", 
    default-features = false 
}
...
```

### Make the Standard Features Available to the Compiler {: #add-standard-features }

> Having declared the module in the workspace `Cargo.toml` file, the dependency can now be added to the specific template `Cargo.toml` file, which for this example that uses the Tanssi GitHub repo, is located in the folder `container-chains/templates/simple/runtime`.

Reviewer 2: ```suggestion
Having declared the module in the workspace `Cargo.toml` file, the dependency can now be added to the specific template `Cargo.toml` file, which, for this example that uses the Tanssi GitHub repo, is located in the folder `container-chains/templates/simple/runtime`.
```

```toml
[dependencies]
...
pallet-toggle = { workspace = true }
...
```

In the same `Cargo.toml` file, add the following features.

```toml
[features]
default = [
	"std",
]
std = [
	...,
	"pallet-toggle/std",
   ...
]
...
runtime-benchmarks = [
	...,
	"pallet-toggle/runtime-benchmarks",
]

try-runtime = [
	...,
	"pallet-toggle/try-runtime",
]
```

### Configure and Add the Module to the Runtime {: #configure-module-in-the-runtime }

> Next, add the following snippet in the `lib.rs` file inside the runtime folder. This configures the module and add the module within the `construct_runtime!`macro.

Reviewer 2: ```suggestion
Next, add the following snippet to the `lib.rs` file inside the runtime folder. This configures the module and adds the module within the `construct_runtime!` macro.
```

```rust
...
impl pallet_template::Config for Runtime {
	type RuntimeEvent = RuntimeEvent;
	type WeightInfo = pallet_template::weights::SubstrateWeight<Runtime>;
}

construct_runtime!(
    pub enum Runtime
    {
        ...
        ...
        Toggle: pallet_toggle,
    }
);
```

### Compile Runtime {: #compile-runtime }

After completing the preceding steps, the module is declared a dependency in the project, configured, and added to the runtime. 

Compile the template using the following command:

```bash
cargo build -p container-chain-template-simple-node --release
```

The terminal output will display an error, similar to the following, caused by different modules referencing different versions of the same dependency:

```bash
error: failed to select a version for `syn`.
```

> ### Patch depedencies {: #patch-dependencies }

Reviewer 2: ```suggestion
### Patch Dependencies {: #patch-dependencies }
```
